### PR TITLE
CDXML agents and reactions support. #832, #836, #837, #835, #834, #832, #830, #853

### PR DESCRIPTION
**Pull request by even1024**
Extended CDXML import/export support. Agents support.

### CODE REVIEW

File: core/indigo-core/molecule/src/base_molecule.cpp [hunk 1]
```diff
@@ -278,37 +279,13 @@ void BaseMolecule::_mergeWithSubmolecule_Sub(BaseMolecule& mol, const Array<int>
...
-        {
-            int bond_idx = findEdgeIndex(mapping[edge.beg], mapping[edge.end]);
-            if (bond_idx > -1)
-            {
-                reaction_bond_reacting_center[bond_idx] = mol.reaction_bond_reacting_center[j];
-            }
-        }
-    }
-
-    // trick for molecules with incorrect stereochemistry, of which we do permutations
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/base_molecule.cpp:302]: Why do we remove this block. Do we have all tests passed?

File: core/indigo-core/molecule/src/molecule_cdxml_loader.cpp [hunk 2]
```diff
@@ -599,20 +634,235 @@ void MoleculeCdxmlLoader::_parseBond(CdxmlBond& bond, const XMLAttribute* pAttr)
...
+        std::string text_element = pTextStyle->Value();
+        if (text_element == "s")
+        {
+            font_face = 0;
+            font_size = 0.0;
+            auto pStyleAttribute = pTextStyle->FirstAttribute();
+            applyDispatcher(pStyleAttribute, style_dispatcher);
+            std::vector<std::string> text_vec_styles;
+            CDXMLFontStyle fs(font_face);
+            if (font_face == 96)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_cdxml_loader.cpp:791]: add constant

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp [hunk 3]
```diff
@@ -33,7 +34,7 @@ IMPL_ERROR(MoleculeCdxmlSaver, "molecule CDXML saver");
 
 MoleculeCdxmlSaver::MoleculeCdxmlSaver(Output& output) : _output(output)
 {
-    _bond_length = BOND_LENGTH;
+    _bond_length = SCALE;
     _max_page_height = 64;
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_cdxml_saver.cpp:38]: add constant

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp [hunk 4]
```diff
@@ -213,13 +214,13 @@ void MoleculeCdxmlSaver::addDefaultFontTable()
 
     name.readString("font", true);
     id = 3;
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_cdxml_saver.cpp:216]: it is hard to understand the logic with fonts. Please add constant for ids

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp [hunk 5]
```diff
@@ -703,6 +703,235 @@ void MoleculeCdxmlSaver::saveMoleculeFragment(BaseMolecule& mol, const Vec2f& of
...
+void MoleculeCdxmlSaver::addMetaData(const MetaDataStorage& meta, int id)
+{
+    const auto& meta_objects = meta.metaData();
+    for (int meta_index = 0; meta_index < meta_objects.size(); ++meta_index)
+    {
+        id++;
+        PropertiesMap attrs;
+        attrs.clear();
+        auto pobj = meta_objects[meta_index];
+        switch (pobj->_class_id)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_cdxml_saver.cpp:715]: it seems no default for the switch

File: utils/indigo-depict/main.c [hunk 6]
```diff
@@ -843,7 +844,7 @@ int main(int argc, char* argv[])
     indigoSetErrorHandler(onError, 0);
 
     indigoSetOption("ignore-stereochemistry-errors", "on");
-    indigoSetOption("molfile-saving-mode", "3000");
+    indigoSetOption("molfile-saving-mode", "2000");
```
Comment (AlexanderSavelyev) [inline utils/indigo-depict/main.c:847]: why did we change to v2000?

File: utils/indigo-depict/main.c [hunk 7]
```diff
@@ -896,7 +901,7 @@ int main(int argc, char* argv[])
         _prepare(obj, p.aromatization);
         if (p.action == ACTION_LAYOUT)
         {
-            indigoLayout(obj);
+            // indigoLayout(obj);
```
Comment (AlexanderSavelyev) [inline utils/indigo-depict/main.c:904]: why layout is disabled?

File: core/indigo-core/molecule/src/molecule_cdxml_loader.cpp (hunk 2 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_cdxml_loader.cpp:790]: fixed

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp (hunk 3 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_cdxml_saver.cpp:38]: fixed

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp (hunk 5 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_cdxml_saver.cpp:715]: Default statement on its place in the bottom of the body of the switch

File: utils/indigo-depict/main.c (hunk 7 above)
Comment (even1024) [reply utils/indigo-depict/main.c]: fixed

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp (hunk 5 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_cdxml_saver.cpp:714]: fixed

**Review by AlexanderSavelyev — APPROVED** [review]

File: core/indigo-core/molecule/src/molecule_cdxml_saver.cpp (hunk 4 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_cdxml_saver.cpp]: fixed

**Pull request merged**